File: Question0/Go/Concurrent/main.go
package main

import (
	"fmt"
	"sync"
)

func productOfArrayExceptSelf(nums []int) []int {
	var leftProd []int
	var rightProd []int = make([]int, len(nums))
	var result []int

	wg := sync.WaitGroup{}

	// [1, 2, 3, 4]
	// right = [24, 12, 4, 1]
	// left = [1, 1, 2, 6]

	// Iterate through the array and calculate both the
	leftProd = append(leftProd, 1) // Initialize the left side first element
	rightProd[len(nums)-1] = 1     // Initialize the right side last element

	wg.Add(2)
	go func() {
		for i := 1; i < len(nums); i++ {
			product := leftProd[i-1] * nums[i-1]
			leftProd = append(leftProd, product)
		}
		wg.Done()
	}()

	go func() {
		for i := len(nums) - 2; i >= 0; i-- {
			product := rightProd[i+1] * nums[i+1]
			rightProd[i] = product
		}
		wg.Done() // Tell the waitgroup this task is finished
	}()
	wg.Wait()

	// Last loop, multiply everything
	for i := 0; i < len(nums); i++ {
		product := leftProd[i] * rightProd[i]
		result = append(result, product)
	}

	return result
}

func main() {
	numbers := []int{-1, 1, 0, -3, 3}
	fmt.Println("Input: ", numbers)
	numbers2 := productOfArrayExceptSelf(numbers)
	fmt.Println("Output: ", numbers2)
}

File: Question0/Python/Concurrent/main.py
import threading

def leftProducts(nums, left) -> list[int]:
    n = len(nums)
    for i in range(1, n):
        left[i] = left[i - 1] * nums[i - 1]
    return left

def rightProducts(nums, right) -> list[int]:
    n = len(nums)
    for i in range(n - 2, -1, -1): # start at n-2, end at 0 (-1), decrement -1
        right[i] = right[i + 1] * nums[i + 1]
    return right

def productOfArrayExceptSelf(nums) -> list[int]:
    n = len(nums)
    left = [1] * n
    right = [1] * n # If n == 3, this creates an array [1, 1, 1]

    thread1 = threading.Thread(target = leftProducts, args = (nums, left))
    thread2 = threading.Thread(target = rightProducts, args = (nums, right))

    thread1.start()
    thread2.start()

    thread1.join()
    thread2.join() # Wait for the thread to join back to the main thread

    result = []
    for i in range(n):
        result.append(left[i] * right[i])

    return result


def main():
    result = productOfArrayExceptSelf([1, 2, 3, 4, 5])
    print(result)

if __name__ == "__main__":
    main()